### PR TITLE
[Operators] LayerNorm Kernels + LigerLayerNorm

### PR DESCRIPTION
**Pull request by AndreSlavescu**
## Summary

integrated layernorm custom kernels + LigerLayerNorm module

## Testing Done

tested layernorm kernels for correctness

- Hardware Type: RTX 3090
- [x] run `make test` to ensure correctness
- [x] run `make checkstyle` to ensure code style
- [x] run `make test-convergence` to ensure convergence

### CODE REVIEW

File: src/liger_kernel/ops/layer_norm.py [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+    except ModuleNotFoundError:
+        # for working with NGC containers
+        from triton.language.extra.cuda.libdevice import rsqrt
+else:
+    from triton.language.math import rsqrt
+
+
+@triton.jit
+def _layer_norm_forward(
+    Y_ptr,
```
Comment (lancerts) [inline src/liger_kernel/ops/layer_norm.py:27]: looks almost good. Can we add some comments on the shape similar to https://github.com/linkedin/Liger-Kernel/pull/145/files#diff-9778ccd57e6d91be7bf3a5a42d621a010150abcbf55ca5e4be9821b914fea6fcR24?

File: src/liger_kernel/ops/layer_norm.py [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+    db_row = tl.zeros((BLOCK_SIZE,), dtype=tl.float32)
+
+    X_ptr += row_start * stride_x
+    Mean_ptr += row_start
+    RSTD_ptr += row_start
+    DX_ptr += row_start * stride_dx
+    DY_ptr += row_start * stride_dy
+
+    for _ in range(row_start, row_end):
+        x = tl.load(X_ptr + cols, mask=mask, other=0.0).to(tl.float32)
```
Comment (lancerts) [inline src/liger_kernel/ops/layer_norm.py:107]: do we need to cast to f32? @ByronHsu @yundai424 , does this slow things down?

File: src/liger_kernel/ops/layer_norm.py [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+    @staticmethod
+    @ensure_contiguous
+    def backward(ctx, dY):
+        shape = dY.shape
+        dim = shape[-1]
+        dY = dY.view(-1, dim)
+        X, W, B, Mean, RSTD = ctx.saved_tensors
+        n_rows, n_cols = dY.shape
+
+        assert X.stride(-1) == 1
```
Comment (lancerts) [inline src/liger_kernel/ops/layer_norm.py:185]: we can remove these if we have ensured the contiguous.

File: src/liger_kernel/ops/layer_norm.py (hunk 2 above)
Comment (AndreSlavescu) [reply src/liger_kernel/ops/layer_norm.py:107]: tested without casts. Works fine + is a bit faster, but marginal difference.

Comment (lancerts) [reply src/liger_kernel/ops/layer_norm.py]: Just wondering, how much faster? ballpark is fine. 5%, 10%, 2%?

**Review by lancerts — APPROVED** [review]
lgtm!

**Pull request merged**